packages/hub/indexing/document-builder.js

```diff
@@ -192,3 +220,17 @@ function ensure(obj, section) {
   }
   return obj.data[section];
 }
+
+function flattenJsonapi(record) {
+  // passthru if it doesn't look like json api
+  if (!record.data) { return record; }
+
+  let { data } = record;
+  let { id, type } = data;
+  let flattenedRecord = { id, type };
+
+  for (let attr of Object.keys(data.attributes || {})) {
+    flattenedRecord[attr] = data.attributes[attr];
+  }
+  return flattenedRecord;
+}
```

Comment:
I don't think this is right -- we shouldn't be guessing which things to flatten by looking at the value. We know precisely which fields should be flattened (because they are relationships) based on the schema.

It also seems weird to me that it only handles attributes. Flattening needs to happen recursively across all levels of relationships too, but if that's implemented elsewhere, I don't see why attribute flattening would be happening separately here.

Reply:
we know which fields are relationships, but some relationships are strings, some are arrays strings, some are dates, some are arrays of objects. this particular function is used to flatten the fields that ultimately become the search doc embellishments. in that scenario, there is no need to flatten a relationship of a relationship--just the attributes will do. In fact I had inadvertently done that in a initial pass thru this and resulted in failing tests, and my query to you around if we should be able to return results for queries that where indirect relationships match query params, and you said not to do that.

Reply:
right now, the search doc embellishments are not json api and we only capture the top level relationships. this is not to be confused with the json api relationships that are used to feed into the includes. that a totally separate thing (which wis recursive and handled correctly as evidenced by the passing tests).

Reply:
and just as nomenclaure, when I say "searchdoc embellishments" i mean all the fields that appear outside of the pristine part of the search doc.

Reply:
All the things that appear inside `included` in the pristine doc also need to appear _inline_ in the search doc, down arbitrarily deep. Top level relationships are not sufficient.

Given pristine:
```
{
  data: {
    type: 'posts',
    id: '1',
    attributes: { title: 'hello world' },
    relationships: {
      author: { 
        data: { type: 'people', id: '1' }
      }
    }
  },
  included: [
    { 
      type: 'people',
      id: '1',
      attributes: { name: 'Hassan' },
      relationships: {
        company: {
          data: { type: 'companies', id: '2' }
        }
      }
   },
   {
      type: 'companies',
      id: '2',
      attributes: { name: 'Cardstack' }
   }
  ]
}
```

We need searchDoc:

```
{
  type: 'posts',
  id: '1',
  title: 'hello world' ,
  author: {
     type: 'people',
     id: '1',
     name: 'Hassan',
     company: {
       type: 'companies',
       id: '2',
       name: 'Cardstack'
     }
  }
}
```

So that a natural search of `author.company.name=Cardstack` will find this document.

Reply:
Ok, we’ll probably have to address some breaking tests for this too, then.

Reply:
This is definitely clear and actionable feedback that I can make headway on.

Reply:
I would have hoped we already have test coverage for deeply nested queries like this, but maybe not.

Reply:
BTW, the searchDoc flattening I showed above is necessarily safe from namespace collisions because JSON:API requires all attributes, relationships, and "type" and "id" to live in a single namespace. That spec is really smart.

Reply:
I’ll make sure to test for that too